# The Battle of Neighborhoods - Final Assignment

In this Final Assignment we will examine the Neighborhoods of Brussels in order to identify the best Neighborhoods for a young couple that recently moved to the Belgian capital. The Real Estate company recently added a additional service that offers their clients a tailored service including a detailed report on the best Neighborhoods based on the preferences of the client.  

### Table of Content

* [1. Introduction: Business Problem](#Introduction)
* [2. Data](#Data)
    * [2.1 Scrapping and Cleaning](#Scrapping_and_Cleaning)
    * [2.2 Latitude and Longitude Data](#Latitude_and_Longitude)
    * [2.3 Visualize the map](#Visualize_Map)
    * [2.4 List of venues](#Foursquare_venues)
* [3. Methodology](#Methodology)
* [4. Results](#Results)
* [5. Discussions](#Discussion)
* [6. Conclusion](#Conclusion)

## 1. Introduction: Business Problem <a class="anchor" id="Introduction"></a>

A young couple recently moved to the Belgian capital for work and are wondering which of the Neighborhoods of Brussels will be the best to live in. Their interests include:  
1. The outdoors
2. Italian Restaurants
3. Cultural activities

They just saw an advertisment for a tailored service from a Real Estate company that will help them identify the top 3 Neighborhoods to choose based on their interests. Therefore, they contacted this Real Estate company and transmitted their interests.  
The young couple both are in their 30s and work full-time and don't know Brussels at all. As they will have to move in the next weeks to the Belgian capital they will need to find a new appartment quickly. 
The Real Estate company told them that they could find them an appartment in two weeks. The first week would be dedicated in coming up with the top 3 Neighborhoods tailored to the couple and the second week would be used to find appartments in these Neighborhoods.

The young couple are quite excited in seeing the analysis of the Real Estate and are curious what kind of Neighborhoods the Real Estate company will come up with. As they don't have any aquaintance in the city they depend 100% from the Real Estate company.

The Real Estate company analyzes their interests and determines that the best way to come up with the best Neighborhoods they will need the coordinates of Brussels and combine this with the Foursquare location data. 

## 2. Data <a class="anchor" id="Data"></a>

In order to determine the best Neighborhoods for the young couple the Real Estate company first analyzes how Brussels is divided.  
Brussels is the capital of Belgium and is divided in 19 municipalities, each whith a defined Postal Code.  
The municipalities can also be divided in different Neighborhoods. However, the Real Estate company finds out that these smaller sections don't have very defined limits and no specific Postal Codes.

Therefore, they decided to restrict their recommendation to one of the 19 municipalities of Brussels.  

Furthermore regarding these findings the Real Estate company decides that the data they will need is:

       - A list of each 19 municipalities of Brussels combined with their postal code and the geographic coordinates.
       - Categories of the venues inside these 19 municipalities. 

1. *List of 19 municipalities*  
For the list of each 19 municipalities of Brussels, the Real Estate company found a Wikipedia page that includes the 19 municipalities with their corresponding postal codes. They will scrap the Wikipedia page to extract the table.  
However, they still need to get the latitude and longitude coordinates for each postal code. For this they found a CSV file from the Open Data platform of Wallonia and Brussels that gives the full list of each postal code in Belgium combined with its respective longitude and latitude data.  

2. *Categories of venues*  
In order to get a list of venues and their categories the Real Estate company decides to use the Foursquare API and create a venue list based on the geographic coordinates identified previously.  

To find a list of the 19 Brussels Municipalities the Real Estate company found the Wikipedia page that has a table that lists the 19 municipalities with their corresponding Postal Codes.

### 2.1 Scrapping and Cleaning the Wikipedia page <a class="anchor" id="Scrapping_and_Cleaning"></a>

To start the company scrapes the Wikipedia page that hosts the Table of the different 19 municipalities of Brussels.  

The first step is to install and import the different modules we will need to scrape and clean the Table. For this operation we will need Pandas and Numpy.

In [1]:
import pandas as pd
import numpy as np

To scrape the Wikipedia page the company decides to use the Pandas read_html method that returns HTML tables into a list of DataFrame objects.  They prefer this method to the BeautifulSoup package as the pandas method is more straightforward for extracting table data from a HTML page.

The URL of the Wikipedia page that contains the Table is: "https://fr.wikipedia.org/wiki/R%C3%A9gion_de_Bruxelles-Capitale"

In [2]:
url = 'https://fr.wikipedia.org/wiki/R%C3%A9gion_de_Bruxelles-Capitale'

df_list = pd.read_html(url)

len(df_list)

16

We can see that read_html identified 16 different HTML tables. From navigating to the Wikipedia page we can identify that the table we are interested in is the second HTML table present. Therefore, let's only look at the second DataFrame object.

In [3]:
brussels_list = df_list[1]

brussels_list.head()

Repèresur lacarte                                            Commune  \
0                1.0                                         Anderlecht   
1                2.0                                          Auderghem   
2                3.0                              Berchem-Sainte-Agathe   
3                4.0  Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                5.0                                          Etterbeek   

                                   Code postal  Superficie(en km2)  \
0                                         1070                1774   
1                                         1160                 903   
2                                         1082                 295   
3  Plusieurs : 1000, 1040, 1050 1020 1120 1130                3261   
4                                         1040                 315   

  Population(au 1/1/2009) Population(au 1/1/2010) Population(au 1/1/2016)  \
0                 101 371                 104 647                 116 494   
1                  30 456                  30 811                  33 085   
2                  21 669                  22 185                  24 176   
3                 153 377                 157 673                 176 512   
4                  43 512                  44 352                  47 023   

  Population(au 1/1/2019)[36] Population(au 1/1/2020)[37] Évolution 2010-2020  \
0                     118 920                     120 009            + 15 362   
1                      33 970                      34 342             + 3 531   
2                      25 195                      25 195             + 3 010   
3                     179 797                     183 287            + 25 614   
4                      48 008                      48 194             + 3 842   

          Δ%  
0  + 14,68 %  
1  + 11,46 %  
2  + 13,56 %  
3  + 16,24 %  
4   + 8,66 %

From the above DataFrame we can see that the table includes a lot of information that we don't need for the moment. Let's only keep the relevant columns, "Commune" and "Code postal" which translates into "municipalities" and "Postal Code". 

In [4]:
brussels_list =  brussels_list [['Commune', 'Code postal']]

brussels_list

Commune  \
0                                          Anderlecht   
1                                           Auderghem   
2                               Berchem-Sainte-Agathe   
3   Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                                           Etterbeek   
5                                               Evere   
6                                              Forest   
7                                           Ganshoren   
8                                             Ixelles   
9                                               Jette   
10                                         Koekelberg   
11                               Molenbeek-Saint-Jean   
12                                       Saint-Gilles   
13                              Saint-Josse-ten-Noode   
14                                         Schaerbeek   
15                                              Uccle   
16                                Watermael-Boitsfort   
17                               Woluwe-Saint-Lambert   
18                                Woluwe-Saint-Pierre   
19                                             Totaux   

                                    Code postal  
0                                          1070  
1                                          1160  
2                                          1082  
3   Plusieurs : 1000, 1040, 1050 1020 1120 1130  
4                                          1040  
5                                          1140  
6                                          1190  
7                                          1083  
8                                          1050  
9                                          1090  
10                                         1081  
11                                         1080  
12                                         1060  
13                                         1210  
14                                         1030  
15                                         1180  
16                                         1170  
17                                         1200  
18                                         1150  
19                                          NaN

The table looks clean. However, we can see that the fourth row has more than one name and more than one postal code. That's because the central municipality of Brussels includes, aside of itself 3 other regions that formerly were seperate municipalities but are now attached to the municipality of Brussels. As they were formerly municipalities these 3 regions have distinct postal codes.  

Therefore, we will add these regions to the complete DataFrame.  

First let's have a look to the fourth row in detail.

In [5]:
city_brussels_PostalCode = brussels_list.iloc[3]

city_brussels_PostalCode

Commune        Ville de Bruxelles : Bruxelles Laeken Neder-Ov...
Code postal          Plusieurs : 1000, 1040, 1050 1020 1120 1130
Name: 3, dtype: object

In [6]:
city_brussels_name = city_brussels_PostalCode['Commune']

print(city_brussels_name)

Ville de Bruxelles : Bruxelles Laeken Neder-Over-Heembeek Haren


We can see that the row includes 4 distinct names. Namely: Bruxelles, Laeken, Neder-Over-Heembeek, and Haren. Therefore, let's create a list that includes these 4 municipalities names. 

In [7]:
city_brussels_name_list = city_brussels_name.split(" ")
city_brussels_name_list

['Ville',
 'de',
 'Bruxelles\xa0:',
 'Bruxelles',
 'Laeken',
 'Neder-Over-Heembeek',
 'Haren']

In [8]:
city_brussels_name_lists = city_brussels_name_list[3:7]

city_brussels_name_lists

['Bruxelles', 'Laeken', 'Neder-Over-Heembeek', 'Haren']

Let's have a look at the postal codes in this row. 

In [9]:
city_brussels_PostalCode_lists = city_brussels_PostalCode['Code postal'].split(" ")

city_brussels_PostalCode_lists

['Plusieurs\xa0:', '1000,', '1040,', '1050', '1020', '1120', '1130']

We can see that there are more postal codes than names. That's because "Bruxelles" had a sub-list of postal codes; 1000, 1040, and 1050. 

In [10]:
city_brussels_PostalCode_lists = city_brussels_PostalCode_lists[1:7]

city_brussels_PostalCode_lists

['1000,', '1040,', '1050', '1020', '1120', '1130']

Upon further inspection we can see that two of these postal codes; 1040 and 1050, are already present in the table. Therefore, we will only keep 1000 for Bruxelles.

In [11]:
del city_brussels_PostalCode_lists[1:3]

The last step is to remove the comma that appears after 1000.

In [12]:
city_brussels_PostalCode_lists = [city_brussels_PostalCode_lists[0].replace('1000,', '1000'), city_brussels_PostalCode_lists[1],city_brussels_PostalCode_lists[2], city_brussels_PostalCode_lists[3]]

city_brussels_PostalCode_lists

['1000', '1020', '1120', '1130']

With both list clean we are ready to merge them and create a sub-DataFrame.

In [13]:
brussels_city_df = pd.DataFrame({'Commune':city_brussels_name_lists, 'Code postal': city_brussels_PostalCode_lists})

brussels_city_df

Commune Code postal
0            Bruxelles        1000
1               Laeken        1020
2  Neder-Over-Heembeek        1120
3                Haren        1130

Now its time to fuse the two DataFrames together.

In [14]:
full_df = brussels_list.append(brussels_city_df, ignore_index=True)

full_df

Commune  \
0                                          Anderlecht   
1                                           Auderghem   
2                               Berchem-Sainte-Agathe   
3   Ville de Bruxelles : Bruxelles Laeken Neder-Ov...   
4                                           Etterbeek   
5                                               Evere   
6                                              Forest   
7                                           Ganshoren   
8                                             Ixelles   
9                                               Jette   
10                                         Koekelberg   
11                               Molenbeek-Saint-Jean   
12                                       Saint-Gilles   
13                              Saint-Josse-ten-Noode   
14                                         Schaerbeek   
15                                              Uccle   
16                                Watermael-Boitsfort   
17                               Woluwe-Saint-Lambert   
18                                Woluwe-Saint-Pierre   
19                                             Totaux   
20                                          Bruxelles   
21                                             Laeken   
22                                Neder-Over-Heembeek   
23                                              Haren   

                                    Code postal  
0                                          1070  
1                                          1160  
2                                          1082  
3   Plusieurs : 1000, 1040, 1050 1020 1120 1130  
4                                          1040  
5                                          1140  
6                                          1190  
7                                          1083  
8                                          1050  
9                                          1090  
10                                         1081  
11                                         1080  
12                                         1060  
13                                         1210  
14                                         1030  
15                                         1180  
16                                         1170  
17                                         1200  
18                                         1150  
19                                          NaN  
20                                         1000  
21                                         1020  
22                                         1120  
23                                         1130

We can see that this Dataframe still includes the 'faulty' row. Furthermore, it includes a Total row that has a *NaN* value. Therefore, let's drop these two rows.

In [15]:
full_df.drop([3, 19], inplace=True)

Lastly, let's sort the DataFrame by increasing number of postal code.

In [16]:
full_df.sort_values(by=['Code postal'], ignore_index=True)

Commune Code postal
0               Bruxelles        1000
1                  Laeken        1020
2              Schaerbeek        1030
3               Etterbeek        1040
4                 Ixelles        1050
5            Saint-Gilles        1060
6              Anderlecht        1070
7    Molenbeek-Saint-Jean        1080
8              Koekelberg        1081
9   Berchem-Sainte-Agathe        1082
10              Ganshoren        1083
11                  Jette        1090
12    Neder-Over-Heembeek        1120
13                  Haren        1130
14                  Evere        1140
15    Woluwe-Saint-Pierre        1150
16              Auderghem        1160
17    Watermael-Boitsfort        1170
18                  Uccle        1180
19                 Forest        1190
20   Woluwe-Saint-Lambert        1200
21  Saint-Josse-ten-Noode        1210

In [17]:
full_df.shape

(22, 2)

We see that the DataFrame now includes 21 postal codes, which corresponds to the 19 municipalities of Brussels plus the two added sub-regions of the Brussels municipality.

In order to get the longitutde and latitude coordinates of the 19 different municipalities, the Real Estate company has access to the Open Data website of Wallonia and Brussels that hosts a CSV file listing all Belgian Postal Codes with the names of the municipalities as well as their longitude and latitudes.


## 2.2 Adding the Latitude and Longitude data to the municipalities. <a class="anchor" id="Latitude_and_Longitude"></a>

In this section we will explore how to add the Latitude and Longitude coordinates to the municipalities using the postal codes as an index. 

The CSV used is openly accessible from the Open Data Wallonie-Bruxelles platform. [Link](https://www.odwb.be/explore/dataset/code-postaux-belge/information/)

As this CSV uses a semicolon to seperate the data we will have to specify this when reading the CSV into a DataFrame.

In [18]:
csv = 'https://www.odwb.be/explore/dataset/code-postaux-belge/download/?format=csv&timezone=Europe/Berlin&lang=fr&use_labels_for_header=true&csv_separator=%3B'

lat_long = pd.read_csv(csv, sep=';')

lat_long.head()

Code             Localite  Longitude   Latitude           Coordonnees  Geom
0  1020               Laeken   4.348713  50.883392   50.883392,4.3487134   NaN
1  1030           Schaerbeek   4.373712  50.867604  50.8676041,4.3737121   NaN
2  1090                Jette   4.326090  50.877763  50.8777634,4.3260903   NaN
3  1140                Evere   4.402160  50.870452  50.8704524,4.4021602   NaN
4  1170  Watermael-Boitsfort   4.415818  50.799394   50.799394,4.4158177   NaN

Like before, we don't need all the columns present in the CSV. The columns that interest us are the Longitude and Latitude. Furthermore, we will need one column to act as a key when we will merge the two DataFrame together. As the postal code is unique the 'Code' column will also have to be kept.

In [19]:
lat_long = lat_long[['Code', 'Latitude', 'Longitude']]

Moreover, in order to use the postal code as an index we will have to rename the column to match the column of the postal code in our **brussels_df** DataFrame. 

In [20]:
lat_long.rename(columns = {'Code': 'Code postal'}, inplace = True)

We are ready to merge the **full_df** DataFrame with the **lat_long** DataFrame on the common column 'Code postal'. 
This will add the latitude and longitude coordinate to each municipality identified previously.

In [21]:
# brussels_df = full_df.merge(lat_long,on=['Code postal'])

The error message indicated that the columns in the two DataFrames 'Code postal' used as a unique key don't have the same datatype. Let's have a look at both DataFrames and identify the data types of each column.

In [22]:
full_df.dtypes

Commune        object
Code postal    object
dtype: object

We can see that the two columns of the **full_df** DataFrame have the data type object. This is correct for the 'Commune' columns as we are speaking about the names of the municipalities. However, the 'Code postal' column should not have the object datatype but a numeric data type.  

Now let's look at the data types of the **lat_long** DataFrame.

In [23]:
lat_long.dtypes

Code postal      int64
Latitude       float64
Longitude      float64
dtype: object

We see that the column 'Code postal' of the **lat_long** DataFrame has the data type *int64*. Therefore, let's convert the 'Code postal' of the **full_df** DataFrame to int64.  
That way can then finally merge both DataFrames. 

In [24]:
full_df = full_df.astype({'Code postal':'int64'})

Let's verify if the data type has been well converted.

In [25]:
full_df.dtypes

Commune        object
Code postal     int64
dtype: object

We can see now that the 'Code postal' column in both DataFrames have the same data type and can be safely merged on this column.

In [26]:
brussels_df = full_df.merge(lat_long,on=['Code postal'])

In [27]:
brussels_df

Commune  Code postal   Latitude  Longitude
0              Anderlecht         1070  50.838141   4.312340
1               Auderghem         1160  50.815657   4.433139
2   Berchem-Sainte-Agathe         1082  50.863984   4.292702
3               Etterbeek         1040  50.836851   4.389510
4                   Evere         1140  50.870452   4.402160
5                  Forest         1190  50.809143   4.317751
6               Ganshoren         1083  50.871240   4.317510
7                 Ixelles         1050  50.822285   4.381571
8                   Jette         1090  50.877763   4.326090
9              Koekelberg         1081  50.862263   4.325708
10   Molenbeek-Saint-Jean         1080  50.854355   4.322778
11           Saint-Gilles         1060  50.826741   4.345668
12  Saint-Josse-ten-Noode         1210  50.853074   4.372336
13             Schaerbeek         1030  50.867604   4.373712
14                  Uccle         1180  50.801820   4.337235
15    Watermael-Boitsfort         1170  50.799394   4.415818
16   Woluwe-Saint-Lambert         1200  50.846693   4.428484
17    Woluwe-Saint-Pierre         1150  50.829243   4.443297
18              Bruxelles         1000  50.846557   4.351697
19                 Laeken         1020  50.883392   4.348713
20    Neder-Over-Heembeek         1120  50.897796   4.390489
21                  Haren         1130  50.891966   4.412571

The brussels_df DataFrame now includes the names of the municipalities, their corresponding postal code, and their latitude and longitude coordinates. This DataFrame can then be used to identify the municipalities on a map.

## 2.3 Visualization of the municipalities on a map <a class="anchor" id="Visualize_Map"></a>

In this section we will visualize the identified municipalities on a Map of Brussels.  

First lets install and import the need packages and modules.

In [28]:
import folium #package used to visualize a map

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import numpy as np # library to handle data in a vectorized manner

print('packages installed and imported')

import json

import requests

packages installed and imported


Let's use the Nominatim geolocator to find out the latitude and longitude of Brussels. 

In [29]:
address = 'Brussels'

geolocator = Nominatim(user_agent="br_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Brussels are: {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Brussels are: 50.8465573, 4.351697.


We'll use Brussels latitude and longtiude to center the Folium map and add markers to the identified municipalities on the map. 

In [30]:
map_brussels = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, municipality in zip(brussels_df['Latitude'], brussels_df['Longitude'], brussels_df['Commune']):
    label = '{}'.format(municipality)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_brussels)  
    
map_brussels

Once we have the latitude and longitude of the identified municipalities of Brussels we can then use Foursquare's location data to identify the type of venues in the different municipalities. The type of venues combined with the frequency of these venues inside of the identified municipalities could give us a good idea of which municipality would match the best the criteria's given by the young couple.

## 2.4 List of venues in the municipalities by Category <a class="anchor" id="Foursquare_venues"></a>

In this section we will make an API call to the Foursquare API in order to get information about the venues that are located within the municipalities we identifed earlier. For this we will create a function called getNearbyVenues and that will read into a new DataFrame the name of the venue combined with the municipality in which it is located and the coordinates of that municipality as well as the coordinates of the venues location and its category. 

First lets define the Foursquare API credentials to make it easier afterwards to recall the needed information.

In [31]:
CLIENT_ID = 'NIDDKXAR0Y1JRGEEWFFQHC234U455BN21PQDFE3EFDRP1GJC' # your Foursquare ID
CLIENT_SECRET = '2FO3JXOSKZJLNJKOYE3L120JFVE1T2S5UKDCJC5NGP2PKVST' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

Secondly, we will define the function called getNearbyVenues that will loop through the names of the municipalities and return data about the venues that are located within these municipalities. The radius was set at 500m and the API limit is set at 100. 

In [32]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Commune', 
                  'Commune Latitude', 
                  'Commune Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now let's make use of this new function and read the **brussels_df** into it.

In [33]:
brussels_venues = getNearbyVenues(names=brussels_df['Commune'], latitudes = brussels_df['Latitude'], longitudes=brussels_df['Longitude'])

Anderlecht
Auderghem
Berchem-Sainte-Agathe
Etterbeek
Evere
Forest
Ganshoren
Ixelles
Jette
Koekelberg
Molenbeek-Saint-Jean
Saint-Gilles


KeyError: 'groups'

Let's have look at the first 5 rows of the venues identified thanks to the Foursquare API call. 

In [ ]:
brussels_venues.head()

Furthermore, let's see how many venues were identified for all the municipalities of Brussels.

In [ ]:
print(brussels_venues.shape)

We see that we identified **684** different venues located in the municipalities of Brussels.  

In the next steps we will analyse these venues and identify the municipalities that include the most venues corresponding to the criterias stated by the young couple.

## 3. Methodology <a class="anchor" id="Methodology"></a>

This section will use the Foursquare API to identify the different types of venues present in the different municipalities. Calculating the frequency of appearance of the different types of venues by municipalities will in turn be used to cluster the municipalities in similar municipalities.  
From there we wil analyze the different clusters in order to identify the clusters that respond the best to the criteria's given by the young couple. 

Let's first identify how many different categories of venues were identified in Brussels. 

In [ ]:
brussels_venues_grouped = pd.DataFrame(brussels_venues.groupby('Commune')['Venue Category'].count())

brussels_venues_grouped.head()

In [ ]:
print('There are {} uniques categories in Brussels.'.format(len(brussels_venues['Venue Category'].unique())))

In order to identify which municipality houses to most of venues relevant for our young couple we will transform the DataFrame to have the categories of venues in columns, which will make it easier afterwards to determine the frequency of each category of venue inside the different municipalities.

In [ ]:
# one hot encoding
brussels_onehot = pd.get_dummies(brussels_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
brussels_onehot['Commune'] = brussels_venues['Commune'] 

# move neighborhood column to the first column
fixed_columns = [brussels_onehot.columns[-1]] + list(brussels_onehot.columns[:-1])
brussels_onehot = brussels_onehot[fixed_columns]

brussels_onehot.head(20)

In [ ]:
brussels_onehot.shape

We can see that we have the **684** different venues listed and the **179** unique venue categories in the columns. Now we will group these venues by municipality and calculate the frequency of the different categories inside of the different municipalities.

In [ ]:
brussels_venues_grouped = brussels_onehot.groupby('Commune').mean().reset_index()

brussels_venues_grouped

Now that we have the frequency of occurence of the different categories of venues grouped by the municipalities of Brussels we can merge this DataFrame with the previous **brussels_df** DataFrame to get the Latitude and Longitude of each municiaplity as well as it's postal code. 

In [ ]:
brussels_venues_grouped_full = brussels_venues_grouped.merge(brussels_df, on = ['Commune'])

brussels_venues_grouped_full

As there are quite some Venue types and the Real Estate company is mainly interested in venues that correspond to the 3 criterias given by the young couple they perform a preliminary analysis taking into account only Italian Restaurants as this is an already present venue category.

In [ ]:
from folium import plugins
from folium.plugins import HeatMap

map_italian_restaurant = folium.Map(location=[latitude, longitude], zoom_start=12) 

# Ensure you're handing it floats
brussels_venues['Venue Latitude'] = brussels_venues['Venue Latitude'].astype(float)
brussels_venues['Venue Longitude'] = brussels_venues['Venue Longitude'].astype(float)

# Filter the DF for columns, then remove NaNs
heat_df = brussels_venues[brussels_venues['Venue Category']=='Italian Restaurant'] # Reducing data size so it runs faster
heat_df = heat_df[['Venue Latitude', 'Venue Longitude']]
heat_df = heat_df.dropna(axis=0, subset=['Venue Latitude','Venue Longitude'])

# List comprehension to make out list of lists
heat_data = [[row['Venue Latitude'],row['Venue Longitude']] for index, row in heat_df.iterrows()]

# Plot it on the map
HeatMap(heat_data).add_to(map_italian_restaurant)

# Display the map
map_italian_restaurant

From the map we can see that there is a concentration of Italian Restaurants in the centre of Brussels in the Commune of Saint Josse, in Saint Gilles, and between the municipalities of Ganshoren and Jette. This can already give us some indication about the ideal municipalities for the young couple. However, we will need to also compare this information with the frequency of the outdoor and cultural venues in the different municipalities. Therefore, we will start by listing the different types of venues we could find using the Foursquare API. 

In [ ]:
print(brussels_venues['Venue Category'].unique())

From this list we can see that several venue categories are related to cultural or outdoor venues. We will first determine which venues are related to cultural activities as well as which are related to outdoor activities. 

From the above list we can see that we have several theaters and musuems as well as Jazz and Comedy clubs. Let's combine these in a list called *matching_culture*. 

Afterwards, we'll do the same for the outdoor venues. From the above list we can see that we have Parks and Playgrounds but also Basketball Courts and Hockey Fields. Let's group these venues under the list called *matching_outdoors*.

To finish off we will put the Italian Restaurants into a seperate list called *matching_IT_Restaurants*. 

In [ ]:
matchers_culture = ['Historic', 'Museum', 'Theater', 'Opera', 'Art', 'Arts', 'Exhibit', 'Concert', 'Music', 'Comedy', 'Jazz']
matching_culture = [s for s in brussels_venues['Venue Category'].unique() if any(xs in s for xs in matchers_culture)]

print(matching_culture)

matchers_outdoor = ['Park', 'Playground', 'Court', 'Field', 'Trail', 'Skating', 'Pitch']
matching_outdoor = [s for s in brussels_venues['Venue Category'].unique() if any(xs in s for xs in matchers_outdoor)]

print(matching_outdoor)

matching_IT_Restaurant =['Italian Restaurant']
print(matching_IT_Restaurant)


Once we finished this we can join these 3 lists into one list that we will call *matching_list* as we will use this list to filter out only the venues that fall into these categories. 

In [ ]:
matching_list = matching_culture + matching_outdoor + matching_IT_Restaurant

print(matching_list)

Now let's slice our **brussels_venues** DataFrame and only consider the relevant venues for our young couple. We will call this new DataFrame **brussels_venues_focused**. 

In [ ]:
brussels_venues_focused = brussels_venues[brussels_venues['Venue Category'].isin(matching_list)]

brussels_venues_focused.head()

We can see that we now have less venues and only venues that correspond to our criteria. 

Now we would like to group the different venue types into only 3 different categories. To do this we will create a new column in the **brussels_venues_focused** DataFrame called *Venue Grouped Category*. For each Venue Category we will assign a parameter called *Culture*, *Outdoor*, or *Italian Restaurant*. 

In [ ]:
conditions = [brussels_venues_focused['Venue Category'].isin(matching_culture), brussels_venues_focused['Venue Category'].isin(matching_outdoor), brussels_venues_focused['Venue Category'].isin(matching_IT_Restaurant)]

choices = ['Culture', 'Outdoor', 'Italian Restaurant']

brussels_venues_focused['Venue Grouped Category'] = np.select(conditions, choices, default=0)

brussels_venues_focused

As before we are more interested in the category of venue and the municipality. So let's transform the DataFame by having the Grouped Venue Categories in the columns. This will help us calculating the number of times a certain category of venue is present in each municipalities.

In [ ]:
# one hot encoding
brussels_onehot_focused = pd.get_dummies(brussels_venues_focused[['Venue Grouped Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
brussels_onehot_focused['Commune'] = brussels_venues_focused['Commune'] 

# move neighborhood column to the first column
fixed_columns_focused = [brussels_onehot_focused.columns[-1]] + list(brussels_onehot_focused.columns[:-1])
brussels_onehot_focused = brussels_onehot_focused[fixed_columns_focused]

brussels_onehot_focused.head(20)

In [ ]:
brussels_onehot_focused.shape

We can see from the shape of the new DataFrame **brussels_onehot_focused** that we are now only considering 92 different venues. Now let us group the different categories of venues by municipality in order to determine which municipality is the most attractive for our young couple.  

In [ ]:
brussels_grouped = brussels_onehot_focused.groupby('Commune').mean().reset_index()

brussels_grouped

From the new DataFrame **brussels_grouped** we can see that some municipalities only have one or two of the categories of venues that the young couple is looking for. Therefore, let's clean out this new DataFrame to only keep the municipalities that have at least all three categories of venues. 

In [ ]:
top_brussels = brussels_grouped[(brussels_grouped != 0).all(1)].sort_values(by='Outdoor', ascending=False)

Our DataFrame is seriously reduced as we are only left with 4 different municipalities. As the young couple is mostly interested in Outdoor activities we sorted the DataFrame in descending order of Outdoor venue frequencies. This shows us that Jette is the muncipality with the most Outdoor venues and is the best municipality for our young couple.

Sorting by Outdoor venues frequency the top 3 municipalities for our young coupled are Jette, Saint-Gilles, and Saint-Josse-ten-Noode, which also corresponded to the heatmap of Italian Restaurants inside Brussels. 

To finish let's add the longitude and latitudes of the municipalities to this DataFrame as well as a column that tells us the rank of each municipality.

In [ ]:
final_df = top_brussels.merge(brussels_df, on = 'Commune')

final_df['Rank'] = final_df.index+1

final_df

## 4. Results <a class="anchor" id="Results"></a>

The Real Estate company now has narrowed down the 3 top municipalities for our young couple and is ready to present them with appartements in these 3 specific municipalities. 

Furthermore, they noted that the results matched the preleminary analysis of Italian Restaurants frequency in the different municipalities, as this was the second most important criteria it was interesting to see that the final result was a 100% match with this analysis. 

As there were only 4 municipalities that actually incorporated the 3 different venues categories asked for by the young couple, the Real Estate company decided to include the 4th municipality of Etterbeek in their search for appartements. 

Finally, they decided to present a map of these 4 municipalities to the young couple for them to have a better idea of where these are situated and to make a better decision based on their needs of commuting to work. 

In [ ]:
final_map = folium.Map(location=[latitude, longitude], zoom_start=12) 

for lat, lng, municipality, rank in zip(final_df['Latitude'], final_df['Longitude'], final_df['Commune'], final_df['Rank']):
    label = '{}, {}'.format(municipality, rank)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(final_map)  
    
final_map

## 5. Discussion <a class="anchor" id="Discussion"></a>

Based on the criterias stated by the young couple the best municipality to find an appartment would be Jette as it is the municipality that has the highest frequency of Outdoor venues and second highest frequency of Cultural venues. However, based on the map, Jette is one of the municiaplities that is the furthest away from the center of Brussels and could therefore pose a problem in terms of commuting.  
Therefore, the second best option, Saint-Gilles, would be a better fit should the young couple not have a car. Depending on the workplace the two last municipalities could be also more advantageous in terms of location. 

## Conclusion <a class="anchor" id="Conclusion"></a>

During this exercise I got to learn more in detail on how to pars information from different sources as well as how to use the Foursquare API to extract interesting location data. Crossing this with pre-determined criterias I was able to conclude on what municipalities were best suited based on the given criterias. 

Furthermore, in order to visually analyze some patterns I could make a more in-depth use of the Folium package.

However, as I started out wanting to do a k-mean clustering I figured through out the exercise that this would not be needed in the end as the set problem already narrowed down the different solutions. Therefore, I decided to skip the clustering of the different municipalities of Brussels. Moreover, when clustering using all the available venues given by Foursquare I was not satisfied with the clustering made as it took into account to many different categories of venues and did not reply to the question at hand. 